src/performance.ts

```diff
@@ -0,0 +1,162 @@
+	 * coverted into a csv to paste in your favorite spreadsheet editor. Results are ordered by their total duration.
+	 *
+	 * @example
+	 * const summary = profiler.getPerformanceSummary();
+	 * const csv = summary.map(item =\>
+	 *     `${item.xpath},${item.times},${item.average},${item.totalDuration}`);
+	 * await navigator.clipboard.write(csv);
+	 *
+	 * @public
+	 *
+	 */
+	getPerformanceSummary(): XPathPerformanceMeasurement[];
+
+	/**
+	 * Set the impormentation of the Performance API object. this should implement the Performance interface.
+	 *
+	 * This is usually either window.performance (in the Browser) or global.performance (for NodeJS)
+	 *
+	 * @public
+	 */
+	setPerformanceImplementation(performance: Performance): void;
+
+	/**
+	 * Start profiling XPaths. All marks are cleared. Use {@link Profiler.stopProfiling} to stop it again.
+	 *
+	 * @public
+	 */
+	startProfiling(): void;
+
+	/**
+	 * Stop profiling XPaths, use the {@link Profiler.getPerformanceSummary} function to get hold of the
+	 * summarized results.
+	 *
+	 * @public
+	 */
+	stopProfiling(): void;
+};
+
+/**
+ * @public
+ */
+export const profiler: Profiler = {
+	getPerformanceSummary() {
+		const xpathEntries = performance.getEntriesByType('measure').filter((entry) => {
+			return entry.name.startsWith('xpath: ');
+		});
+
+		return Array.from(
+			xpathEntries
+				.reduce((summedMeasurements, measurement) => {
+					const xpath = measurement.name.substring('xpath: '.length);
+					if (summedMeasurements.has(xpath)) {
+						const summedMeasurement = summedMeasurements.get(xpath);
+						summedMeasurement.times++;
+						summedMeasurement.totalDuration += measurement.duration;
+					} else {
+						summedMeasurements[xpath] = {
+							xpath,
+							times: 1,
+							totalDuration: measurement.duration,
+						};
+					}
+
+					return summedMeasurements;
+				}, new Map<string, XPathPerformanceMeasurement>())
+				.values()
+		)
+			.map((entry) => {
+				entry.average = entry.totalDuration / entry.times;
+				return entry;
+			})
+			.sort((entryA, entryB) => {
+				return entryB.totalDuration - entryA.totalDuration;
+			});
+	},
+
+	setPerformanceImplementation(newPerformanceImplementation: Performance) {
+		performance = newPerformanceImplementation;
+	},
+
+	startProfiling() {
+		if (performance === null) {
+			throw new Error(
+				'Performance API object must be set using `profiler.setPerformanceImplementation` before starting to profile'
+			);
+		}
+		performance.clearMarks();
+		profilingEnabled = true;
+	},
+
+	stopProfiling() {
+		profilingEnabled = false;
+	},
+};
+
+export function markXPathStart(xpath: string) {
+	if (!profilingEnabled) {
+		return;
+	}
+	performance.mark(xpath);
+}
+export function markXPathEnd(xpath: string) {
+	if (!profilingEnabled) {
+		return;
+	}
+	// Replace the mark with a measure of the time spent
+	performance.measure(`xpath: ${xpath}`, xpath, undefined);
+	performance.clearMarks(xpath);
```

Comment:
Will this break if I call the same xpath function recursively twice? For this I could register a custom xpath function which keeps executing itself until an external value changes.

Reply:
Fixed by adding a counter